### PR TITLE
fix: bugs discovered during simulation

### PR DESCRIPTION
**Pull request by zsystm**
**Fixes**
* add missing genesis state `RedelegationInfo`
* remove del shares checking in ChunkInvariantsCheck
  * shares can be decreased (e.g. Slashing during ReDelegation)
* fix wrong implementation of RedelegationInfo.Matured
  * it was reversed. must check based on currTime.
  * at some time, currTime will pass the `rinfo.CompletionTime`, then it should return `true` but formal implementation will return `false`.
* fix RePairRankedInsurances logic
  * outIns from restOutInsurances can be  Paired.
  * fix https://github.com/b-harvest/Canto/issues/28
* fix https://github.com/b-harvest/Canto/issues/26
* fix https://github.com/b-harvest/Canto/issues/27
  * when do `mustPairChunkAndDelegate`, must use latest Validator state.
* fix buggy sim operations (e.g. SimulateMsgProvideInsurance)

**Gas Optimization**
* separate InsuranceState which will be used when query states of module

**SPEC**
* add description for `InsuranceState`

**Simulation**
* implement operations of lsm
* return noOps in expected error situations
* applied PowerReduction `10^6`
* applied static genesis and params (same with Mainnet) 
* update weights (more `LiquidUnstake` and `ClaimDiscountedReward`)
* (temp) applied PowerReduction `10^6` when calc epoch mint provision

**Dependency**
* (temp) replace cosmos-sdk to https://github.com/b-harvest/cosmos-sdk/releases/tag/v0.45.9-canto-lsm-sim to run lsm simulation successfully.

**Refactoring**
* move advance epoch logic to app module level

## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] targeted the correct branch (see [PR Targeting](https://github.com/canto-network/canto/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] included the necessary unit and integration tests
- [x] reviewed "Files changed" and left comments if necessary
 
### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: app/sim_test.go [hunk 1]
```diff
@@ -308,6 +315,7 @@ func TestAppStateDeterminism(t *testing.T) {
 }
 
 func TestAppSimulationAfterImport(t *testing.T) {
+	liquidstakingtypes.ChunkSize = sdk.TokensFromConsensusPower(250_000, sdk.DefaultPowerReduction)
```
Comment (dongsam) [inline app/sim_test.go:318]: This part can be removed because it is already being done in `init()`

File: app/sim_test.go [hunk 2]
```diff
@@ -210,9 +214,12 @@ func TestAppImportExport(t *testing.T) {
 		{app.keys[ibctransfertypes.StoreKey], newApp.keys[ibctransfertypes.StoreKey], [][]byte{}},
 		// {app.keys[evmtypes.StoreKey], newApp.keys[evmtypes.StoreKey], [][]byte{}},
 		{app.keys[feemarkettypes.StoreKey], newApp.keys[feemarkettypes.StoreKey], [][]byte{}},
-		//{app.keys[inflationtypes.StoreKey], newApp.keys[inflationtypes.StoreKey], [][]byte{}},
+		{app.keys[inflationtypes.StoreKey], newApp.keys[inflationtypes.StoreKey], [][]byte{}},
 		// {app.keys[erc20types.StoreKey], newApp.keys[erc20types.StoreKey], [][]byte{}},
-		//{app.keys[epochstypes.StoreKey], newApp.keys[epochstypes.StoreKey], [][]byte{}},
+		{app.keys[epochstypes.StoreKey], newApp.keys[epochstypes.StoreKey], [][]byte{
```
Comment (dongsam) [inline app/sim_test.go:219]: It should be commented, [Ref](https://github.com/Canto-Network/Canto/blob/main/app/sim_test.go#L218-L219)

File: app/params/weights.go [hunk 3]
```diff
@@ -1,16 +1,17 @@
...
 	DefaultWeightMsgCancelProvideInsurance      int = 10
 	DefaultWeightMsgDepositInsurance            int = 10
 	DefaultWeightMsgWithdrawInsurance           int = 20
 	DefaultWeightMsgWithdrawInsuranceCommission int = 10
-	DefaultWeightMsgClaimDiscountedReward       int = 10
+	DefaultWeightMsgClaimDiscountedReward       int = 20
 
 	DefaultWeightUpdateDynamicFeeRateProposal int = 2
 	DefaultWeightUpdateMaximumDiscountRate        = 2
+	DefaultWeightAdvanceEpoch                 int = 5
```
Comment (dongsam) [inline app/params/weights.go:16]: Unused weight variables

File: app/app.go [hunk 4]
```diff
@@ -414,9 +412,10 @@ func NewCanto(
 	app.AccountKeeper = authkeeper.NewAccountKeeper(
 		appCodec, keys[authtypes.StoreKey], app.GetSubspace(authtypes.ModuleName), ethermint.ProtoAccount, maccPerms,
 	)
-	app.BankKeeper = bankkeeper.NewBaseKeeper(
+	baseKeeper := bankkeeper.NewBaseKeeper(
 		appCodec, keys[banktypes.StoreKey], app.AccountKeeper, app.GetSubspace(banktypes.ModuleName), app.BlockedAddrs(),
 	)
+	app.BankKeeper = baseKeeper
```
Comment (dongsam) [inline app/app.go:418]: Is there a reason or intention to do this?

File: x/inflation/types/inflation_calculation.go [hunk 5]
```diff
@@ -42,6 +40,6 @@ func CalculateEpochMintProvision(
 	// Multiply epochMintProvision with power reduction (10^18 for evmos) as the
 	// calculation is based on `evmos` and the issued tokens need to be given in
 	// `aevmos`
-	epochProvision = epochProvision.Mul(ethermint.PowerReduction.ToDec())
+	epochProvision = epochProvision.Mul(sdk.DefaultPowerReduction.ToDec())
```
Comment (dongsam) [inline x/inflation/types/inflation_calculation.go:43]: This part needs to be reverted before the PR merges

File: x/liquidstaking/keeper/grpc_query.go [hunk 6]
```diff
@@ -323,5 +323,11 @@ func (k Keeper) States(c context.Context, req *types.QueryStatesRequest) (*types
...
 	}
 	ctx := sdk.UnwrapSDKContext(c)
-	return &types.QueryStatesResponse{NetAmountState: k.GetNetAmountState(ctx)}, nil
+	nas := k.GetNetAmountState(ctx)
+	is := k.GetInsuranceState(ctx)
+
+	return &types.QueryStatesResponse{
+		NetAmountState: nas,
+		InsuranceState: is,
+	}, nil
```
[the remark this answers is not in the document]
Comment (zsystm) [reply x/liquidstaking/keeper/grpc_query.go:332]: Thanks for the feedback. Frontend team already applied formal endpoints [in here](https://github.com/Plex-Engineer/liquid-staking/blob/main/app/actions.ts#L228-L260).

How about creating a PR related with this change for them? I think. I can do it and need no much time.

[the remark this answers is not in the document]
Comment (dongsam) [reply x/liquidstaking/keeper/grpc_query.go:332]: I don't think it's necessary to change the response struct due to refactor for internal logic, it's better to update it at once when future modifications are inevitable

If they are already using it, use the internal getter function that has been refactored, but as a result, 'QueryStatesResponse' would be better to use the existing structure as it is

File: x/liquidstaking/types/redelegation_info_test.go [hunk 7]
```diff
@@ -41,8 +41,8 @@ func (suite *redelegationInfoTestSuite) TestMatured() {
 	c := types.NewChunk(1)
 	blockTime := time.Now()
 	// sub one hour from blockTime
-	oneHourBeforeBlockTime := blockTime.Add(-time.Hour)
-	info := types.NewRedelegationInfo(c.Id, oneHourBeforeBlockTime)
+	oneHourAfterBlockTime := blockTime.Add(time.Hour)
+	info := types.NewRedelegationInfo(c.Id, oneHourAfterBlockTime)
```
[the remark this answers is not in the document]
Comment (zsystm) [reply x/liquidstaking/types/redelegation_info_test.go:45]: updated it. https://github.com/b-harvest/Canto/pull/31/commits/5b6759dd8287e1435ffb6bdfa9b36930d119b5d6

File: app/params/weights.go (hunk 3 above)
Comment (zsystm) [reply app/params/weights.go]: done in https://github.com/b-harvest/Canto/commit/a9b210366eed46d0c3a9baaaa2a580bb8199b8eb

File: app/app.go (hunk 4 above)
Comment (zsystm) [reply app/app.go]: I think we can delete it. Done in https://github.com/b-harvest/Canto/commit/781a833530b010dc56fa834b259d284468ba2632.

File: app/sim_test.go (hunk 2 above)
Comment (zsystm) [reply app/sim_test.go]: done in https://github.com/Canto-Network/Canto/blob/main/app/sim_test.go#L218-L219 

comment is written by myself but will use comment you linked when merge main branch.

File: x/liquidstaking/keeper/grpc_query.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zsystm) [reply x/liquidstaking/keeper/grpc_query.go]: I added `NetAmountStateEssentials` which is used by all core logics of lsm.
NetAmount is used just for query.

Changes: https://github.com/b-harvest/Canto/compare/b-harvest:e47ac8%5E%5E...b-harvest:e47ac8

**Comment by dongsam** [comment]
In order to solve the power reduction issue I mentioned yesterday, I think the reduction modification in canto should be reverted to ethermint and revised from sdk simulation to big int

**Comment by zsystm** [comment]
> In order to solve the power reduction issue I mentioned yesterday, I think the reduction modification in canto should be reverted to ethermint and revised from sdk simulation to big int

Yeah definitely we need to do that. I started that work after requesting your review for NetAmountStateEssential part.

During your review for updated codes, I'll handle that.

File: x/liquidstaking/keeper/grpc_query.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zsystm) [reply x/liquidstaking/keeper/grpc_query.go]: @dongsam 
If you are ok with changes, I'll update the spec too. SPEC is not updated yet.

[the remark this answers is not in the document]
Comment (dongsam) [reply x/liquidstaking/keeper/grpc_query.go]: Yeah, that part is looks good. Let's continue with the power reduction part

**Comment by zsystm** [comment]
@dongsam 
Applied ethermint reduction! https://github.com/b-harvest/Canto/pull/31/commits/bed5d904aed6682b9487789137d1cc0e46a3e380

For this, I used b-harvest ibc-go and cosmos-sdk.

File: x/liquidstaking/keeper/grpc_query.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (zsystm) [reply x/liquidstaking/keeper/grpc_query.go]: Updated spec in here https://github.com/b-harvest/Canto/pull/31/commits/04d03d3fedce64a1bf108edc66e151c86a00e856

**Comment by dongsam** [comment]
> @dongsam 
> 
> Applied ethermint reduction! https://github.com/b-harvest/Canto/pull/31/commits/bed5d904aed6682b9487789137d1cc0e46a3e380
> 
> 
> 
> For this, I used b-harvest ibc-go and cosmos-sdk.

Could you check the failed workflow?

**Comment by zsystm** [comment]
> 

Sure! I fixed it. Thanks!

**Review by dongsam — APPROVED** [review]
LGTM, let's proceed additionally with the gas optimization, NetAmountState structure, and customizing sdk simulation parts in separate issues after upstream rebase/merge

**Pull request merged**